src/ast/irbuilderbpf.cpp

```diff
@@ -701,6 +701,125 @@ void IRBuilderBPF::CreateForEachMapElem(Value *ctx,
+
+  // createMapLookup  returns an u8*
+  auto *cast = CreatePointerCast(call, getInt64Ty(), "cast");
+
+  Value *prev_value = CREATE_ATOMIC_RMW(AtomicRMWInst::BinOp::Xchg,
+                                        cast,
+                                        getInt64(1),
+                                        8,
+                                        AtomicOrdering::SequentiallyConsistent);
+
+  Function *set_parent = GetInsertBlock()->getParent();
+  BasicBlock *value_is_set_block = BasicBlock::Create(module_.getContext(),
+                                                      "value_is_set",
+                                                      set_parent);
+  Value *set_condition = CreateICmpEQ(prev_value,
+                                      getInt64(0),
+                                      "value_set_condition");
+  CreateCondBr(set_condition, merge_block, value_is_set_block);
+
+  SetInsertPoint(value_is_set_block);
+  /*
+   * The counter is set, we need to exit early from the probe.
+   * This should only happen for the functions that can lead
+   * to a recursive crash e.g. "queued_spin_lock_slowpath"
+   */
+  CreateRet(getInt64(0));
+
+  SetInsertPoint(lookup_failure_block);
+
+  CreateDebugOutput(
+      "Value for per-cpu map key 0 is null. This shouldn't happen.",
+      std::vector<Value *>{},
+      loc);
+  CreateRet(getInt64(0));
+
+  SetInsertPoint(merge_block);
+}
+
+void IRBuilderBPF::CreateUnSetRecursion(const location &loc)
+{
+  const std::string map_ident = to_string(MapType::RecursionPrevention);
+
+  AllocaInst *key = CreateAllocaBPF(getInt32Ty(), "lookup_key");
+  CreateStore(getInt32(0), key);
+
+  CallInst *call = createMapLookup(map_ident, key);
+
+  Function *parent = GetInsertBlock()->getParent();
+  BasicBlock *lookup_success_block = BasicBlock::Create(module_.getContext(),
+                                                        "lookup_success",
+                                                        parent);
+  BasicBlock *lookup_failure_block = BasicBlock::Create(module_.getContext(),
+                                                        "lookup_failure",
+                                                        parent);
+  BasicBlock *merge_block = BasicBlock::Create(module_.getContext(),
+                                               "lookup_merge",
+                                               parent);
+
+  // Make the verifier happy with a null check even though the value should
+  // never be null for key 0.
+  Value *condition = CreateICmpNE(
+      CreateIntCast(call, GET_PTR_TY(), true),
+      ConstantExpr::getCast(Instruction::IntToPtr, getInt64(0), GET_PTR_TY()),
+      "map_lookup_cond");
+  CreateCondBr(condition, lookup_success_block, lookup_failure_block);
+
+  SetInsertPoint(lookup_success_block);
+
+  CreateLifetimeEnd(key);
+
+  // createMapLookup  returns an u8*
+  auto *cast = CreatePointerCast(call, getInt64Ty(), "cast");
+  CreateStore(getInt64(0), cast);
```

Comment:
Does this need to be an atomic store?

Reply:
I asked @4ast the same question above. His response: "The final ST val, 0 is fine, since on entry we got exclusive access. ST val, 0 is WRITE_ONCE equivalent. JITs must not tear the store."

Reply:
Ha, shoulda read closer. Interesting, my naive understanding is that you cannot depend on that for other archs (other than x86-64). :shrug: